Src/Plugins/Security/SEC101_040.ShopifySharedSecretValidator.cs

```diff
@@ -17,41 +17,36 @@ static ShopifySharedSecretValidator()
             Instance = new ShopifySharedSecretValidator();
         }
 
-        public static ValidationState IsValidStatic(ref string matchedPattern,
-                                                    ref Dictionary<string, string> groups,
-                                                    ref string message,
-                                                    out ResultLevelKind resultLevelKind,
-                                                    out Fingerprint fingerprint)
+        public static IEnumerable<ValidationResult> IsValidStatic(ref string matchedPattern,
+                                                                  ref Dictionary<string, string> groups,
+                                                                  ref string message)
         {
             return IsValidStatic(Instance,
                                  ref matchedPattern,
                                  ref groups,
-                                 ref message,
-                                 out resultLevelKind,
-                                 out fingerprint);
+                                 ref message);
         }
 
-        protected override ValidationState IsValidStaticHelper(ref string matchedPattern,
-                                                               ref Dictionary<string, string> groups,
-                                                               ref string message,
-                                                               out ResultLevelKind resultLevelKind,
-                                                               out Fingerprint fingerprint)
+        protected override IEnumerable<ValidationResult> IsValidStaticHelper(ref string matchedPattern,
+                                                                             ref Dictionary<string, string> groups,
+                                                                             ref string message)
         {
-            fingerprint = default;
-            resultLevelKind = default;
-
             if (!groups.TryGetNonEmptyValue("refine", out string secret))
             {
-                return ValidationState.NoMatch;
+                return ValidationResult.NoMatch;
```

Comment:
>ValidationResult [](http://example.com/codeflow?start=23&length=16)

this is clever but you have a problem, i.e., even though ValidationResult.NoMatch is read-only, it's still mutable! so if anyone makes a mistake and changes NoMatch to have an entirely different fingerprint, etc., that change will flow to everyone else.

There are many ways to resolve this. The easiest is to make ValidationResult a struct with no property setters.

A second possibility is to use an interface for the return value. Then you can build a IValidationResult implementing class that is immutable after construction.

Another idea is to use a builder pattern of some kind:

return ValidationResult.CreateNoMatch();

Reply:
Follow on, because fingerprint is a reference class, the struct idea might not be sufficient to easily guarantee immutability. i.e., a struct that provides a reference type by a getter still leaves you vulnerable to changes to that reference type

Reply:
for now, i will go with the third option.
and if needed we can do another pr. this is getting too complex to manipulate :p